### PR TITLE
New engagement date filters [MAILPOET-4989]

### PR DESCRIPTION
**Pull request by johnolek**
## Description

These filters rely on the new operators added in https://github.com/mailpoet/mailpoet/pull/5049, so that should be merged first, and then I will update the base branch for this PR. Feel free to ignore this PR until that one is merged.

This PR adds new dynamic segment filters for the engagement data we're now tracking on the subscriber table. This includes:

- last engagement date
- last purchase date
- last open date
- last click date
- last page view date
- last sending date

## Code review notes

Since these filters should all operate the same as `SubscriberSubscribedDate`, I removed that class in favor of a more general `SubscriberDateField`.

For the tests, I opted for using different fields for the different operators. I could have written tests for every combination of action/operator, but that seemed excessive, and I couldn't immediately think of an elegant way to handle it with a `dataProvider` method. I would welcome other ideas for how to handle the tests.

## QA notes

Please make sure that you have a segment with a  "subscribed date" filter before you start testing and ensure it continues to work the same after installing this version.

## Linked PRs

https://github.com/mailpoet/mailpoet/pull/5049 - Must be merged before this one
https://github.com/mailpoet/mailpoet-premium/pull/749 - should be merged after this one

## Linked tickets

https://mailpoet.atlassian.net/browse/MAILPOET-4989

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

### CODE REVIEW

**Comment by johnolek** [comment]
@rodrigoprimo This PR is dependent on #5049 (based on its branch), so if you'd prefer to wait until that's merged to review this feel free to unassign yourself, but I don't suspect there will be changes to the other PR so I think this should be ok to review before the merge if you have time.

File: mailpoet/lib/Segments/DynamicSegments/Filters/SubscriberDateField.php
Comment (lysyjan) [inline mailpoet/lib/Segments/DynamicSegments/Filters/SubscriberDateField.php:18]: Maybe we can comment on this constant that it has a different pattern name due to backward compatibility. What do you think

File: mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php
Comment (lysyjan) [inline mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php]: Can I ask why did you change the non-static public method to static?
I'm asking because `DateFilterHelper` is a service that should be injected via DI, and in my opinion, those services shouldn't contain static methods.

File: mailpoet/tests/integration/Segments/DynamicSegments/Filters/SubscriberDateFieldTest.php
Comment (lysyjan) [inline mailpoet/tests/integration/Segments/DynamicSegments/Filters/SubscriberDateFieldTest.php:132]: > For the tests, I opted for using different fields for the different operators. I could have written tests for every combination of action/operator, but that seemed excessive, and I couldn't immediately think of an elegant way to handle it with a dataProvider method. I would welcome other ideas for how to handle the tests.

I thought about it and your approach is something I would do. I was also considering something more complex, but it could be very complex, which means easier bugs in the code.

**Review by lysyjan — COMMENTED** [review]
Great job! I like your solution.

I have a question about the change to the static method and a small comment.

File: mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php
Comment (johnolek) [reply mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php]: Thank you for bringing this up. In the back of my mind I've been thinking about updating a lot of the validation code in `FilterDataMapper` to rely on the filter classes themselves, because we end up having to perform most of the same checks in both places. I imagined the constructor growing very large if we needed to inject every single filter that performs any validation, which is why the static method seemed nicer to me.

That being said, I totally agree with you and I will change this back to a public method and inject it into FilterDataMapper.

File: mailpoet/lib/Segments/DynamicSegments/Filters/SubscriberDateField.php
Comment (johnolek) [reply mailpoet/lib/Segments/DynamicSegments/Filters/SubscriberDateField.php:18]: Good idea! I've added it in 169bf537c3ef2044366cc324e846f44df12d7e15

File: mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php
Comment (johnolek) [reply mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php]: I updated the code in 823fdb331d31a7d6aec7635e144fdd7a14cd61ac

For the constructor of `FilterDataMapper`, would it be better to inject the container wrapper instead, or is it ok how I did it?

Comment (lysyjan) [reply mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php]: @johnolek I guess we should avoid using `ContainerWrapper::getInstance()` when it's not necessary or is a good reason for that. Do you think that there is a reason for it? I remember we use it sometimes in temporary code, especially in places that should be removed after finishing Doctrine refactoring.

Comment (johnolek) [reply mailpoet/lib/Segments/DynamicSegments/Filters/DateFilterHelper.php]: No good reason. I’ve seen it in enough places that I thought it might be idiomatic in some cases. I’ll update this to use injection as well.

**Comment by johnolek** [comment]
@lysyjan I pushed some cleanup commits and I think this should be ready to go whenever you have a chance

**Review by lysyjan — APPROVED** [review]
Great job!
Thank you for your patience and cooperation 👍

**Review by Aschepikov — APPROVED** [review]
GJ!

**Pull request merged**
